Add beginnings of unit test coverage of `align`

elebow opened the pull request:
### Description of proposed changes
This PR is intended to serve as an RFC on future unit testing strategy. I'm a software developer with substantial experience in Python development but minimal (essentially layperson) awareness of computational genomics. I would love to contribute to the project, and unit testing seems to be an area where I can help.

A robust unit test suite enables faster development of application code and more confidence in the results it produces. Augur currently has 13% test coverage.

This PR adds coverage of two methods in `align.py`, and a convenience script to ease running of the test suite. If this pattern is acceptable, I would be pleased for it to serve as a model for future unit test coverage. This PR alone increases the test coverage of `align.py` from 12% to 22%, though it was low-hanging fruit.

### Related issue(s)  
none

### Testing
:smiley:

tolot27 commented:
Unit testing would be really helpful. The issue #462 would have been discovered with unit testing. I'd love to see this PR merged soon. Then I can extend with tests for a PR I'll create soon.

danielsoneg commented:
+1 to adding more tests!

One quick suggestion - I've found it useful in the past to make the unit tests as granular as possible - in other words, instead of just `test_check_duplicates`, we'd do `test_check_duplicates_no_dupes`, `test_check_duplicates_string`, `test_check_duplicates_seq`. That both makes it explicit from the tests what the expected behavior under each condition is and also makes it easier to spot the exact error that the tests are throwing.

Also, pytest has a bunch of really nice capabilities, especially [fixtures](https://docs.pytest.org/en/latest/fixture.html) and [tmp_path](https://docs.pytest.org/en/latest/tmpdir.html), that I think could be really useful for this project.

huddlej commented on run_tests.sh:
```diff
@@ -0,0 +1,16 @@
+	coverage_arg='--cov=augur'
+fi
+
+python3 -m pytest -s $coverage_arg "$@"
```
I would prefer to keep all pytest configuration settings in one place. The pytest config options (`-s` and `--cov=augur`) can be moved into the existing pytest config file and always executed. Then this run_tests script could be a single line that runs `python -m pytest -c pytest.python3.ini`.

elebow replied:
:+1: on moving `-s`.

See comment below about omitting `--cov` when `-k` is specified.

huddlej commented on run_tests.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+case "$@" in
+	*-k*) partial_test=1 ;;
```
I couldn't get this flag to work, but maybe just because I don't understand what it is supposed to do. Could you include an example usage with this flag?

elebow replied:
`-k` is passed through to pytest. Pytest will run only tests that match the specified string. https://docs.pytest.org/en/latest/usage.html#specifying-tests-selecting-tests

When using this argument, the coverage output is not useful—Since most tests are skipped, the coverage is incorrect and the output is just noise.

I like to omit `--cov` when running a subset of tests to keep the output easy to read. It fits my test-driven development workflow, but perhaps it's just a personal quirk. I have no objection to moving `--cov` to `pytest.python3.ini` if that is the team's preference, but just want to point out that it makes TDD less convenient.

elebow commented:
On doctest vs standalone tests:

I feel that doctests are most useful as documentation, rather than verification. They're excellent as usage examples, but they can quickly become large and detract from readability of the code. For example, the test code in this PR for `check_duplicates` is 51 lines, and that's not even a complex function.

Doctests also make it hard to use pytest facilities like fixtures, and many editors don't do syntax highlighting of code in comments.

There's little downside to having _both_ doctests and standalone tests, so my preference is usually:
  1. Write doctests primarily as usage examples. Include the happy paths, but don't allow them to grow overly large in an effort to achieve complete verification.
  1. Write standalone tests for complete coverage of all possible (reasonable) cases.

elebow commented:
> One quick suggestion - I've found it useful in the past to make the unit tests as granular as possible - in other words, instead of just `test_check_duplicates`, we'd do `test_check_duplicates_no_dupes`, `test_check_duplicates_string`, `test_check_duplicates_seq`. 

Sounds good to me. I've done it both ways in the past.

danielsoneg commented:
I'm a rando off the street as well, so I'll defer to the maintainers, but given a bit of experience with large/long-lasting projects -

I'd agree with @elebow on unit tests vs doctests. I think the doctests are great, and they've been useful to me as examples, but they're definitely limited, especially around setup and teardown. The `strip_non_reference` tests are leaving a file hanging around because I couldn't find a comfortable way to clean up afterward.

Re --cov, I think my inclination would be to put the "complete" set of test-running flags into the pytest.python3.ini file. I already run a limited set ad-hoc while developing, but I think I'd consider that an "advanced user" feature - for anyone just dropping in out of the blue, the standard "here's how you run tests" docs should run the full suite sufficient to convince us all that new code is good.

elebow commented:
Regarding fixtures:

I left them out of this PR to make for a more incremental changeset to introduce some testing patterns. I'm fully :+1: on using pytest fixtures, and possibly pytest-factoryboy in a later PR. I'm also happy to include them in this PR if that is preferred.

huddlej commented:
Ok, this all sounds excellent and I'm super excited we're having this conversation! @elebow and @danielsoneg, I hope you don't mind if I collect your comments into an improved "Testing" section of our augur dev docs. :)

I see how the `-k` flag and optional running of test coverage provides the advanced user functionality that simply running pytest with defaults wouldn’t.

@elebow, can you make the following small changes, then, and I’ll update the testing docs and Travis CI config accordingly?

- [ ] Move -s into `pytest.python3.ini`
- [ ] Add `pytest-cov` to `setup.py` in the dev dependencies with an appropriate version
- [ ] Update `pytest`’s version in `setup.py` to the latest (`"pytest >=3.2.1, ==3.*",` -> `"pytest >=5.4.1, ==5.*",`)

Then in a separate PR I will:

- [ ] Update the “testing” section of the dev docs to reflect this discussion
- [ ] Update Travis CI config to run `run_tests.sh`
- [ ] Rename `pytest.python3.ini` to the default `pytest.ini`, to save some typing

Then I think we can start accepting more unit tests built along the same approach you've used here.

elebow commented:
I was already working on some changes; I hope this is okay for this PR:

* moved `-s` into `pytest.python3.ini`
* added line breaks and comments to `pytest.python3.ini`
* `run_tests.sh` now includes `-c pytest.python3.ini`, so the complete test suite can be run with just `./run_tests.sh`
* added `pytest-cov` to dev dependencies
* bumped version of `pytest` to `5.4.*`
* split the `check_duplicates` monolithic test function into a set of smaller ones

Push coming in a few minutes.

---

Regarding having Travis CI run `run_tests.sh`:

That's probably fine, but it might be better to have Travis continue to invoke `pytest -c pytest.python3.ini`.

`run_tests.sh` provides a convenience for humans, and we might want to add things there in the future that are not appropriate for Travis—for example, `--color=always`. It's probably fine for Travis to run the script, though.

hameerabbasi commented on run_tests.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
```
I would suggest removing this file and instead adding `--no-cov` when needed to `pytest`.

hameerabbasi commented on pytest.python3.ini:
```diff
@@ -1,2 +1,10 @@
+
+	# run any doctests found in .py modules
+	--doctest-modules
+
```
Add `--cov` here, `--no-cov` can override it.

hameerabbasi commented on pytest.python3.ini:
```diff
@@ -1,2 +1,10 @@
+	# run any doctests found in .py modules
+	--doctest-modules
+
+	augur/
```
Add these to `testpaths`, see here: https://github.com/pydata/sparse/blob/master/pytest.ini#L5-L6

willingc commented:
@huddlej @elebow @danielsoneg Suggestions by @hameerabbasi about Python testing are good practice in the Python world. While doctests are nice and can enhance docs, I would consider them separate from test files using pytest. Running pytest with coverage in CI is good practice. If the tests take a long time to run, Azure Pipelines has been very good for execution speed of tests.

One way to increase coverage would be to determine which source files need tests and add those to an issue with a checklist and folks can leave a message on the issue to claim work on a file and check off the box when PRs are merged.

elebow commented:
:+1: thanks for the feedback.

* Added `--cov=augur` to `pytest.python3.ini`, so it runs even in CI.
* Changed `run_tests.sh` to add `--no-cov` when `-k` is specified. Note that this causes a warning when running the script; see https://github.com/pytest-dev/pytest-cov/issues/284.
* Left `run_tests.sh` in the PR because it still has value as convenience for human users when doing TDD.
* Moved the test paths to from `addopts` to  `testpaths` attribute in the ini.

Changes pushed.

huddlej commented:
This works for me! Thank you everyone for all of your help!

@willingc Would you be interested in taking lead on the issue creation for specific modules to test that you mentioned? I am also happy to do this, but I'll definitely be a bottleneck as I'm working from home with a small child. :)

willingc commented:
Sure thing @huddlej. No problem as I am on vacation (or staycation) this week. I'll add the issue.
